cpp/src/arrow/compute/exec/aggregate.cc

```diff
@@ -101,136 +106,401 @@ Result<FieldVector> ResolveKernels(
+      if (!argument_types.empty()) {
+        // Construct and initialize HashAggregateKernels
+        ARROW_ASSIGN_OR_RAISE(kernels, GetKernels(ctx, aggregates, argument_types));
 
-    states.resize(task_group->parallelism());
-    for (auto& state : states) {
-      ARROW_ASSIGN_OR_RAISE(state, InitKernels(kernels, ctx, aggregates, argument_types));
-    }
+        states.resize(task_group->parallelism());
+        for (auto& state : states) {
+          ARROW_ASSIGN_OR_RAISE(state,
+                                InitKernels(kernels, ctx, aggregates, argument_types));
+        }
 
-    ARROW_ASSIGN_OR_RAISE(
-        out_fields, ResolveKernels(aggregates, kernels, states[0], ctx, argument_types));
+        ARROW_ASSIGN_OR_RAISE(out_fields, ResolveKernels(aggregates, kernels, states[0],
+                                                         ctx, argument_types));
+      } else {
+        out_fields = {};
+      }
+      out_fields.insert(out_fields.end(), key_fields.begin(), key_fields.end());
 
-    RETURN_NOT_OK(argument_iterator.Init(args_batch, ctx->exec_chunksize()));
-  }
+      return Status::OK();
+    }
 
-  // Construct Groupers
-  ARROW_ASSIGN_OR_RAISE(ExecBatch keys_batch, ExecBatch::Make(keys));
-  auto key_types = keys_batch.GetTypes();
+    Status Consume(const BatchInfo& batch_info) {
+      const std::vector<TypeHolder>& argument_types = process.argument_types;
+      ExecContext* ctx = process.ctx;
 
-  std::vector<std::unique_ptr<Grouper>> groupers(task_group->parallelism());
-  for (auto& grouper : groupers) {
-    ARROW_ASSIGN_OR_RAISE(grouper, Grouper::Make(key_types, ctx));
-  }
+      const ExecBatch& args_batch = batch_info.args_batch;
+      const ExecBatch& keys_batch = batch_info.keys_batch;
+      const ExecBatch& segment_keys_batch = batch_info.segment_keys_batch;
+
+      if (segment_keys_batch.length == 0) {
+        return Status::OK();
+      }
+      segment_keys = {};
+      for (auto value : segment_keys_batch.values) {
+        if (value.is_scalar()) {
+          segment_keys.push_back(value.scalar());
+        } else if (value.is_array()) {
+          ARROW_ASSIGN_OR_RAISE(auto scalar, value.make_array()->GetScalar(0));
+          segment_keys.push_back(scalar);
+        } else if (value.is_chunked_array()) {
+          ARROW_ASSIGN_OR_RAISE(auto scalar, value.chunked_array()->GetScalar(0));
+          segment_keys.push_back(scalar);
+        } else {
+          return Status::Invalid("consuming an invalid segment key type ", *value.type());
+        }
+      }
 
-  std::mutex mutex;
-  std::unordered_map<std::thread::id, size_t> thread_ids;
+      if (!argument_types.empty()) {
+        ARROW_RETURN_NOT_OK(argument_iterator.Init(args_batch, ctx->exec_chunksize()));
+      }
+      ARROW_RETURN_NOT_OK(key_iterator.Init(keys_batch, ctx->exec_chunksize()));
+
+      std::mutex mutex;
+      std::unordered_map<std::thread::id, size_t> thread_ids;
```

Comment:
We have `ThreadIndexer` already.  Would that work?  Alternatively, it might be nice / simpler to have something like https://github.com/apache/arrow/pull/12871 if you want to pick that work up and finish it.

Reply:
Looks like `ThreadIndexer` should work. I'll look into this and #12871 later.

Reply:
I'll go with `ThreadIndexer` and defer on #12871.